src/capwap_loc_provider.erl

```diff
@@ -0,0 +1,154 @@
+%% (at your option) any later version.
+
+%% This program is distributed in the hope that it will be useful,
+%% but WITHOUT ANY WARRANTY; without even the implied warranty of
+%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+%% GNU Affero General Public License for more details.
+
+%% You should have received a copy of the GNU Affero General Public License
+%% along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+-module(capwap_loc_provider).
+
+-behavior(gen_server).
+
+%% API
+-export([start/1, start_link/1]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2]).
+
+%% user api
+-export([flush_loc_cache/0, get_loc/1, get_loc/2, load_config/1]).
+
+-include_lib("kernel/include/logger.hrl").
+
+-define(CACHE_LOC, capwap_loc_provider_cache).
+
+-record(loc_cache, {loc_tab, config_tab, refresh, providers, provider_chain = undefined}).
+
+%% callbacks
+
+-callback config_fun(any()) -> fun((<<>>) -> {error, any()} | {location, <<>>, <<>>}).
+%% A separate callback returning an opaque ref would be
+%% useful in the case of stateful providers.
+%% This option is not covered yet.
+
+%% API
+
+start_link(Config) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, Config, []).
+
+start(Config) ->
+    gen_server:start({local, ?MODULE}, ?MODULE, Config, []).
+
+%% validation
+
+validate(#{refresh := RefreshTime, providers := Providers})
+  when is_integer(RefreshTime) andalso RefreshTime>0 ->
+    validate_providers(Providers),
+    ok.
+
+validate_providers(Providers) when is_list(Providers) ->
+    lists:map(fun validate_provider/1, Providers),
+    ok.
+
+validate_provider({Mod, Config}) ->
+    Mod:validate_provider(Config).
+
+%% gen_server callbacks
+
+init(Config = #{refresh := RefreshTime, providers := Providers}) ->
+    validate(Config),
+    ?LOG(debug, "table ref: ~p", [ets:whereis(?CACHE_LOC)]),
+    TabOpts = [set, protected, named_table, {read_concurrency, true}],
+    %% Not needed since it will return the table name
+    LocTab = ets:new(?CACHE_LOC, TabOpts),
+    ?LOG(debug, "cache started: ~p", [LocTab]),
+    ?LOG(debug, "table ref: ~p", [ets:whereis(?CACHE_LOC)]),
+    %% TODO Handle exceptions in chain?
+    {ok, #loc_cache{
+	loc_tab = LocTab, refresh = RefreshTime, providers = Providers, provider_chain = chain(Providers)}}.
+
+handle_call(flush_loc_cache, _, State = #loc_cache{loc_tab = LocTab}) ->
+    ?LOG(debug, "Flush location cache, tableref: ~p", [ets:whereis(LocTab)]),
+    ets:delete_all_objects(LocTab),
+    {reply, ok, State};
+handle_call({get_loc, UseCache = true, Name}, _, State = #loc_cache{
+	     loc_tab = LocTab, refresh = RefreshTime, provider_chain = Chain}) ->
+    ?LOG(debug, "get location for device name (with cache): ~p", [Name]),
+    Result = case ets:lookup(LocTab, Name) of
+	[{Name, Loc = {location, _, _}}] -> Loc;
+	[] -> refresh_loc(Name, Chain, LocTab, RefreshTime)
```

Comment:
the indention is strange, it should be relative to the `case` statement, not relative to the beginning of the line.